### DISCUSSION TITLE
How does CF Tool submit to Codeforces?

### DISCUSSION THREAD

**Discussion by qinyihao** [cpeditor/cpeditor#764] [Q&A] [answered]
I'm also making a contest helper for codeforces, and I would like to ask how can you do the submission and how is its security? Thanks.

**Comment by qinyihao** [comment]
I mean, how do you assure the password is safe and it won't be compromised during the transmission?

**Comment by ouuan** [comment, marked as answer]
Please ask [CF Tool](https://github.com/xalanq/cf-tool/).

CP Editor only executes CF Tool. I don't know the details of how CF Tool submits, but I believe it's safe.

BTW, I don't suggest you reinvent the wheel. Besides CF Tool, there are also [online-judge-tools](https://github.com/online-judge-tools/oj) and [CP Tools](https://github.com/cp-tools/cpt) which are actively maintained. You can have a look at them.

**Reply by qinyihao** [reply]
> Please ask CF Tool.
> 
> CP Editor only executes CF Tool. I don't know the details of how CF Tool submits, but I believe it's safe.
> 
> BTW, I don't suggest you reinvent the wheel. Besides CF Tool, there are also [online-judge-tools](https://github.com/online-judge-tools/oj) and [CP Tools](https://github.com/cp-tools/cpt) which are actively maintained. You can have a look at them.

OK thanks

**Comment by qinyihao** [comment]
@ouuan but I think the creator of CF tool has stopped the maintenance of CF tool, as the last commit was made on Mar1, 2020

**Reply by ouuan** [reply]
That's why I mentioned OJ Tools and CP Tools. We plan to support [custom submit commands](https://github.com/cpeditor/cpeditor/issues/212) in the future.

**Comment by coder3101** [comment]
The password is kept encrypted in your local file system. During login the user name and password are decrypted and sent to login page, and login page saves a cookie to remember this user. This cookie is stored in your local file system. When you submit to CF, it checks if you are logged in by checking if you have that cookie. 

Onces the cookie has expired, you request new cookies by again authenticating on login page. 

This is not complete, I was navigating CF sources and got to know all these. For more information please ask its authors.

**Reply by qinyihao** [reply]
all right thanks!
